src/datasets/features/features.py

```diff
@@ -337,9 +337,9 @@ def _cast_to_python_objects(obj: Any, only_1d_for_numpy: bool, optimize_list_cas
         return obj.values.tolist(), True
     elif isinstance(obj, pd.DataFrame):
         return obj.to_dict("list"), True
-    elif isinstance(obj, dict):
+    elif isinstance(obj, Mapping):
+        has_changed = not isinstance(obj, dict)
```

Comment:
Can you add a comment to explain why we need this ? I'm not sure I understand why it's related to using `{**inputs, **processed_inputs}`

Reply:
I think it's easier to understand this with an example. Let's say we are running the following map:
```python
dset.map(lambda ex: {"translation": ex})
```
In this case `inputs` is equal to `ex` and `processed_inputs` to `{"translation": ex}`. Running `update` in the next step, as we are currently doing, creates an object with infinite recursion:
```python
>>> inputs = {"en": "aa", "fr": "bb"}
>>> processed_inputs = {"translation": inputs}
>>> inputs.update(processed_inputs)
>>> inputs
{'en': 'aa', 'fr': 'bb', 'translation': {...}} <- infinite recursion
```
This PR avoids this with the dict unpacking syntax:
```python
>>> inputs = {"en": "aa", "fr": "bb"}
>>> processed_inputs = {"translation": inputs}
>>> {**inputs, **processed_inputs}
{'en': 'aa', 'fr': 'bb', 'translation': {'en': 'aa', 'fr': 'bb'}}
```
However, that's still not enough for scenarios where we turn the input dictionary (`ex` in my example) into the LazyDict object, a subclass of `UserDict`, and return it as a dict value (`translation` value in my example) because PyArrow only supports the built-in `dict` type hence the above check.